### PR TITLE
Add insert_tcp_options() nasl function

### PR DESCRIPTION
**Pull request by jjnicola**
**What**:
This add support for inserting MSS, Wscale, sack permit and timestamp
tcp options.

**Why**:
It was not supported and was required.

**How**:
Run the nasl script below, and check with tcpdump or wireshark that the options are set as expected.

`openvas-nasl -X -d -i /<install-prefix>/var/lib/openvas/plugins -t <IPV4 or IPv6 to test the funcs> this_script.nasl`
```
if(description) {
  script_oid("1.2.3");
  exit(0);
}

include("misc_func.inc");

myhost = this_host();
port = 22022;
srcport= 35000;
dstip = get_host_ip();
filter = string("tcp and src ", dstip, " and dst ", myhost, " and dst port ", srcport, " and src port ", port);

ip = forge_ip_packet(ip_v   : 4,
                     ip_hl  : 5,
                     ip_tos : 0,
                     ip_len : 20,
                     ip_id  : rand(),
                     ip_p   : IPPROTO_TCP,
                     ip_ttl : 255,
                     ip_off : 0,
                     ip_src : myhost,
                     ip_dst: dstip);


tcp = forge_tcp_packet(ip       : ip,
                       th_ack   : 0,
                       th_dport : port,
                       th_flags : TH_SYN,
                       #th_seq   : tcp_seq + 1024,
                       th_sport : srcport,
                       th_x2    : 0,
                       th_off   : 5,
                       th_win   : 1024,
                       th_urp   : 0,
                       data:"Hola mundo");

display("TCP options:");
display("TCPOPT_SACK_PERMITTED: ", TCPOPT_SACK_PERMITTED);
display("TCPOPT_MAXSEG: ", TCPOPT_MAXSEG);        
display("TCPOPT_WINDOW: ", TCPOPT_WINDOW);        
display("TCPOPT_TIMESTAMP: ", TCPOPT_TIMESTAMP);    

tcp = insert_tcp_options(tcp:tcp,
                                TCPOPT_MAXSEG, 1024,
                                TCPOPT_WINDOW, 14 ,
                                TCPOPT_SACK_PERMITTED,
                                TCPOPT_TIMESTAMP, 123456, 987654);

r = send_packet(tcp, pcap_active:TRUE, pcap_filter:filter);

## IPv6 Check
IP6_v = 0x06;
IP6_P = IPPROTO_TCP;
IP6_HLIM = 0x40;

ip6_packet = forge_ipv6_packet( ip6_v:IP6_v,
                                ip6_p:IP6_P,
                                ip6_hlim:IP6_HLIM,
                                ip6_src:myhost,
                                ip6_dst:dstip );
                                
tcp6 = forge_tcp_v6_packet(ip6 : ip6_packet,
                           th_dport : port,
                           th_flags : TH_SYN,
                           th_sport : srcport,
                           th_x2    : 0,
                           th_win   : 1024,
                           data: "hola mundo");

tcp6 = insert_tcp_v6_options(tcp:tcp6, TCPOPT_WINDOW, 14, TCPOPT_MAXSEG, 1500, TCPOPT_TIMESTAMP, 123456, 2100000);

r = send_v6packet(tcp6, pcap_active:TRUE, pcap_filter:filter);

```

**Checklist**:

- [ ] Tests
- [x] [CHANGELOG](https://github.com/greenbone/openvas/blob/master/CHANGELOG.md) Entry

### CODE REVIEW

File: nasl/nasl_packet_forgery.c [hunk 1]
```diff
@@ -818,6 +847,269 @@ set_tcp_elements (lex_ctxt *lexic)
...
+        break;
+
+      switch (tcp_opt)
+        {
+        case TCPOPT_MAXSEG:
+          tcp_opt_val = get_int_var_by_num (lexic, i + 1, -1);
+          i++;
+          if (tcp_opt_val < (int) TCP_MSS_DEFAULT || tcp_opt_val > 65535)
+            {
+              nasl_perror (lexic, "%s: Invalid value for TCP option MSS'\n",
```
Comment (cfi-gb) [inline nasl/nasl_packet_forgery.c:915]:
```suggestion
              nasl_perror (lexic, "%s: Invalid value for TCP option MSS\n",
```

File: nasl/nasl_packet_forgery.c [hunk 2]
```diff
@@ -818,6 +847,269 @@ set_tcp_elements (lex_ctxt *lexic)
...
+                  sizeof (struct tcp_opt_mss));
+          ptr_opts_pos = ptr_opts_pos + sizeof (struct tcp_opt_mss);
+          g_free (opt_mss);
+          break;
+        case TCPOPT_WINDOW:
+          tcp_opt_val = get_int_var_by_num (lexic, i + 1, -1);
+          i++;
+          if (tcp_opt_val < 0 || tcp_opt_val > 14)
+            {
+              nasl_perror (lexic, "%s: Invalid value for TCP option WScale'\n",
```
Comment (cfi-gb) [inline nasl/nasl_packet_forgery.c:943]:
```suggestion
              nasl_perror (lexic, "%s: Invalid value for TCP option WScale\n",
```

File: nasl/nasl_packet_forgery.c [hunk 3]
```diff
@@ -818,6 +847,269 @@ set_tcp_elements (lex_ctxt *lexic)
...
+                  sizeof (struct tcp_opt_sack_perm));
+          ptr_opts_pos = ptr_opts_pos + sizeof (struct tcp_opt_sack_perm);
+          g_free (opt_sack_perm);
+          break;
+        case TCPOPT_TIMESTAMP:
+          tcp_opt_val = get_int_var_by_num (lexic, i + 1, -1);
+          tcp_opt_val2 = get_int_var_by_num (lexic, i + 2, -1);
+          i = i + 2;
+          if (tcp_opt_val < 0)
+            nasl_perror (lexic, "%s: Invalid value for TCP option Timestamp'\n",
```
Comment (cfi-gb) [inline nasl/nasl_packet_forgery.c:990]:
```suggestion
            nasl_perror (lexic, "%s: Invalid value for TCP option Timestamp\n",
```

File: nasl/nasl_packet_forgery_v6.c [hunk 4]
```diff
@@ -766,6 +795,254 @@ set_tcp_v6_elements (lex_ctxt *lexic)
...
+        break;
+
+      switch (tcp_opt)
+        {
+        case TCPOPT_MAXSEG:
+          tcp_opt_val = get_int_var_by_num (lexic, i + 1, -1);
+          i++;
+          if (tcp_opt_val < (int) TCP_MSS_DESIRED || tcp_opt_val > 65535)
+            {
+              nasl_perror (lexic, "%s: Invalid value for TCP option MSS'\n",
```
Comment (cfi-gb) [inline nasl/nasl_packet_forgery_v6.c:862]:
```suggestion
              nasl_perror (lexic, "%s: Invalid value for TCP option MSS\n",
```

File: nasl/nasl_packet_forgery_v6.c [hunk 5]
```diff
@@ -766,6 +795,254 @@ set_tcp_v6_elements (lex_ctxt *lexic)
...
+                  sizeof (struct tcp_opt_mss));
+          ptr_opts_pos = ptr_opts_pos + sizeof (struct tcp_opt_mss);
+          g_free (opt_mss);
+          break;
+        case TCPOPT_WINDOW:
+          tcp_opt_val = get_int_var_by_num (lexic, i + 1, -1);
+          i++;
+          if (tcp_opt_val < 0 || tcp_opt_val > 14)
+            {
+              nasl_perror (lexic, "%s: Invalid value for TCP option WScale'\n",
```
Comment (cfi-gb) [inline nasl/nasl_packet_forgery_v6.c:890]:
```suggestion
              nasl_perror (lexic, "%s: Invalid value for TCP option WScale\n",
```

File: nasl/nasl_packet_forgery_v6.c [hunk 6]
```diff
@@ -766,6 +795,254 @@ set_tcp_v6_elements (lex_ctxt *lexic)
...
+                  sizeof (struct tcp_opt_sack_perm));
+          ptr_opts_pos = ptr_opts_pos + sizeof (struct tcp_opt_sack_perm);
+          g_free (opt_sack_perm);
+          break;
+        case TCPOPT_TIMESTAMP:
+          tcp_opt_val = get_int_var_by_num (lexic, i + 1, -1);
+          tcp_opt_val2 = get_int_var_by_num (lexic, i + 2, -1);
+          i = i + 2;
+          if (tcp_opt_val < 0)
+            nasl_perror (lexic, "%s: Invalid value for TCP option Timestamp'\n",
```
Comment (cfi-gb) [inline nasl/nasl_packet_forgery_v6.c:937]:
```suggestion
            nasl_perror (lexic, "%s: Invalid value for TCP option Timestamp\n",
```

File: nasl/nasl_packet_forgery.c [hunk 7]
```diff
@@ -818,6 +847,269 @@ set_tcp_elements (lex_ctxt *lexic)
   return retc;
 }
 
+/**
+ * @brief Add options to a TCP segment header.
+ * Possible options are:
+ *   TCPOPT_MAXSEG (2), values between 536 and 65535
+ *   TCPOPT_WINDOW (3), with values between 0 and 14
+ *   TCPOPT_SACK_PERMITTED (4), no value requiered.
```
Comment (cfi-gb) [inline nasl/nasl_packet_forgery.c:855]:
```suggestion
 *   TCPOPT_SACK_PERMITTED (4), no value required.
```

File: nasl/nasl_packet_forgery_v6.c [hunk 8]
```diff
@@ -766,6 +795,254 @@ set_tcp_v6_elements (lex_ctxt *lexic)
   return retc;
 }
 
+/**
+ * @brief Add options to a TCP segment header.
+ * Possible options are:
+ *   TCPOPT_MAXSEG (2), values between 536 and 65535
+ *   TCPOPT_WINDOW (3), with values between 0 and 14
+ *   TCPOPT_SACK_PERMITTED (4), no value requiered.
```
Comment (cfi-gb) [inline nasl/nasl_packet_forgery_v6.c:803]:
```suggestion
 *   TCPOPT_SACK_PERMITTED (4), no value required.
```

**Review by cfi-gb — CHANGES_REQUESTED** [review]

File: nasl/nasl_packet_forgery.c [hunk 9]
```diff
@@ -818,6 +847,269 @@ set_tcp_elements (lex_ctxt *lexic)
...
+          ptr_opts_pos++;
+        }
+    }
+  // Add EOL
+  memcpy (ptr_opts_pos, &eol, 1);
+
+  if (ip->ip_hl * 4 > pktsz)
+    tcp =
+      (struct tcphdr *) (pkt
+                         + 20); /* ip->ip_hl is bogus, we work around that */
```
Comment (ArnoStiefvater) [inline nasl/nasl_packet_forgery.c:1048]: The formatting is bad here. Maybe put the comment above the line so the formatter does not do this.

File: nasl/nasl_packet_forgery.c [hunk 10]
```diff
@@ -818,6 +847,269 @@ set_tcp_elements (lex_ctxt *lexic)
...
+  else
+    tcp = (struct tcphdr *) (pkt + ip->ip_hl * 4);
+
+  if (pktsz < UNFIX (ip->ip_len))
+    return NULL;
```
Comment (ArnoStiefvater) [inline nasl/nasl_packet_forgery.c:1053]: opts need to be freed here I guess.

File: nasl/nasl_packet_forgery.c [hunk 11]
```diff
@@ -818,6 +847,269 @@ set_tcp_elements (lex_ctxt *lexic)
...
+  // Calculate tcp header with options checksum
+  struct pseudohdr pseudoheader;
+  char *tcpsumdata =
+    g_malloc0 (sizeof (struct pseudohdr) + opt_size_allocated + data_len + 1);
+  struct in_addr source, dest;
+
+  source.s_addr = ip->ip_src.s_addr;
+  dest.s_addr = ip->ip_dst.s_addr;
+
+  bzero (&pseudoheader, sizeof (pseudoheader));
```
Comment (ArnoStiefvater) [inline nasl/nasl_packet_forgery.c:1088]: bzero and bcopy are deprecated. memset and memcpy should be used instead.

Comment (ArnoStiefvater) [inline nasl/nasl_packet_forgery.c:1088]: bzero and bcopy are deprecated. memset and memcpy should be used instead. Same for all following occurrences.

**Review by ArnoStiefvater — APPROVED** [review]
Tested with the supplied scripts and tcpdump. Everything works as expected.

**Review by cfi-gb — APPROVED** [review]

**Comment by cfi-gb** [comment]
Reviewed the fixes to my review and approved them (only them, not the whole code changes here which i can't approve).

**Pull request merged**